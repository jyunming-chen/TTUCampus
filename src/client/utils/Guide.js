// @ts-check

import Path from './Path';

/**
 * @typedef {Object} RawPath
 * @prop {string} from
 * @prop {string} to
 * @prop {string[]} points
 */

/**
 * @typedef {Object} RawPoint
 * @prop {string} id
 * @prop {string} name
 * @prop {number[]} position
 */

export default class Guide {
  /**
   * @param {RawPath[]} paths
   * @param {RawPoint[]} points
   */
  constructor(paths, points) {
    /**
     * @type {Record<string, Record<string, Path>>}
     */
    this.map = {};

    /**
     * @type {Record<string, RawPoint>}
     */
    this.points = points.reduce(
      (currentPoints, rawPoint) =>
        Object.assign(currentPoints, { [rawPoint.id]: rawPoint }),
      {},
    );

    this.paths = paths;

    paths.forEach(path => {
      const pathPoints = path.points.map(point => this.points[point].position);
      this.setPath(path.from, path.to, pathPoints);
      this.setPath(path.to, path.from, pathPoints.slice().reverse());
    });
  }

  /**
   * @param {string} from
   * @param {string} to
   * @param {number[][]} points
   */
  setPath(from, to, points) {
    if (!this.map[from]) {
      this.map[from] = {};
    }
    this.map[from][to] = new Path(points);
  }

  /**
   * @param {string} from
   * @param {string} to
   */
  getPath(from, to) {
    if (!this.map[from]) {
      return undefined;
    }
    return this.map[from][to];
  }
}
