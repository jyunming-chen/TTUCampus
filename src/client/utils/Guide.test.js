import Guide from './Guide';

const paths = new Guide(
  [
    {
      from: 'A',
      to: 'B',
      points: ['a', 'b', 'c'],
    },
    {
      from: 'A',
      to: 'C',
      points: ['a', 'd', 'e'],
    },
  ],
  [
    {
      id: 'a',
      name: 'a',
      position: [0, 0],
    },
    {
      id: 'b',
      name: 'b',
      position: [1, 1],
    },
    {
      id: 'c',
      name: 'c',
      position: [2, 2],
    },
    {
      id: 'd',
      name: 'd',
      position: [-1, -2],
    },
    {
      id: 'e',
      name: 'e',
      position: [-3, -4],
    },
  ],
);

test('new Paths(configPaths)', () => {
  expect(paths).toMatchSnapshot();
});

test('#getPath()', () => {
  expect(paths.getPath('unknown', 'unknown')).toBeUndefined();
  expect(paths.getPath('A', 'B')).toMatchSnapshot();
});
