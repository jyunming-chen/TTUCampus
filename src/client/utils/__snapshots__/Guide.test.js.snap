// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getPath() 1`] = `
Path {
  "distance": 2.8284271247461903,
  "distances": Array [
    0,
    1.4142135623730951,
    2.8284271247461903,
  ],
  "points": Array [
    Vector3 {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    Vector3 {
      "x": 1,
      "y": 0,
      "z": -1,
    },
    Vector3 {
      "x": 2,
      "y": 0,
      "z": -2,
    },
  ],
}
`;

exports[`new Paths(configPaths) 1`] = `
Guide {
  "map": Object {
    "A": Object {
      "B": Path {
        "distance": 2.8284271247461903,
        "distances": Array [
          0,
          1.4142135623730951,
          2.8284271247461903,
        ],
        "points": Array [
          Vector3 {
            "x": 0,
            "y": 0,
            "z": 0,
          },
          Vector3 {
            "x": 1,
            "y": 0,
            "z": -1,
          },
          Vector3 {
            "x": 2,
            "y": 0,
            "z": -2,
          },
        ],
      },
      "C": Path {
        "distance": 5.06449510224598,
        "distances": Array [
          0,
          2.23606797749979,
          5.06449510224598,
        ],
        "points": Array [
          Vector3 {
            "x": 0,
            "y": 0,
            "z": 0,
          },
          Vector3 {
            "x": -1,
            "y": 0,
            "z": 2,
          },
          Vector3 {
            "x": -3,
            "y": 0,
            "z": 4,
          },
        ],
      },
    },
    "B": Object {
      "A": Path {
        "distance": 2.8284271247461903,
        "distances": Array [
          0,
          1.4142135623730951,
          2.8284271247461903,
        ],
        "points": Array [
          Vector3 {
            "x": 2,
            "y": 0,
            "z": -2,
          },
          Vector3 {
            "x": 1,
            "y": 0,
            "z": -1,
          },
          Vector3 {
            "x": 0,
            "y": 0,
            "z": 0,
          },
        ],
      },
    },
    "C": Object {
      "A": Path {
        "distance": 5.06449510224598,
        "distances": Array [
          0,
          2.8284271247461903,
          5.06449510224598,
        ],
        "points": Array [
          Vector3 {
            "x": -3,
            "y": 0,
            "z": 4,
          },
          Vector3 {
            "x": -1,
            "y": 0,
            "z": 2,
          },
          Vector3 {
            "x": 0,
            "y": 0,
            "z": 0,
          },
        ],
      },
    },
  },
  "paths": Array [
    Object {
      "from": "A",
      "points": Array [
        "a",
        "b",
        "c",
      ],
      "to": "B",
    },
    Object {
      "from": "A",
      "points": Array [
        "a",
        "d",
        "e",
      ],
      "to": "C",
    },
  ],
  "points": Object {
    "a": Object {
      "id": "a",
      "name": "a",
      "position": Array [
        0,
        0,
      ],
    },
    "b": Object {
      "id": "b",
      "name": "b",
      "position": Array [
        1,
        1,
      ],
    },
    "c": Object {
      "id": "c",
      "name": "c",
      "position": Array [
        2,
        2,
      ],
    },
    "d": Object {
      "id": "d",
      "name": "d",
      "position": Array [
        -1,
        -2,
      ],
    },
    "e": Object {
      "id": "e",
      "name": "e",
      "position": Array [
        -3,
        -4,
      ],
    },
  },
}
`;
